src/types/redis_string.cc

```diff
@@ -374,25 +374,23 @@ rocksdb::Status String::MSet(const std::vector<StringPair> &pairs, uint64_t ttl,
   // Data race, key string maybe overwrite by other key while didn't lock the key here,
   // to improve the set performance
   std::string ns_key;
+  auto batch = storage_->GetWriteBatchBase();
+  WriteBatchLogData log_data(kRedisString);
+  batch->PutLogData(log_data.Encode());
   for (const auto &pair : pairs) {
     std::string bytes;
     Metadata metadata(kRedisString, false);
     metadata.expire = expire;
     metadata.Encode(&bytes);
     bytes.append(pair.value.data(), pair.value.size());
-    auto batch = storage_->GetWriteBatchBase();
-    WriteBatchLogData log_data(kRedisString);
-    batch->PutLogData(log_data.Encode());
     AppendNamespacePrefix(pair.key, &ns_key);
     batch->Put(metadata_cf_handle_, ns_key, bytes);
     std::optional<LockGuard> guard;
```

Comment:
@enjoy-binbin We need to use the multi-key lock before the write operation instead of inside the loop. Because the for loop only inserts key-values into the write batch and it won't affect the DB before the write operation.

Reply:
thanks, i overlook the comment, and just doing a plain merge to update the branch.